Add support for forein keys in schema generation within aggregates

kobaeugenea opened the pull request:
Done:
- [x] Generate column and foreign key for @MapedCollection annotated fields
- [x] Generate only foreign key for @MapedCollection annotated fields if column exists in child entry
- [x] Drop foreign keys
- [ ] Foreign key name generator (hardcoded now)

Always drop foreign keys before table drop. Always add foreign keys after table creation. The same for fields. (to be sore that referenced column always exists).

Closes #1599
Related tickets #756, #1600

schauder reviewed (changes requested):
This looks pretty good already.

I see just one major problem: It won't work for nested lists or maps.
If you could add a test for those and fix the resulting issues it would be ready to go.

kobaeugenea commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/mapping/schema/ForeignKey.java:
```diff
@@ -0,0 +1,27 @@
+ * @author Evgenii Koba
+ * @since 3.2
+ */
+record ForeignKey(String name, String tableName, String columnName, String referencedTableName,
```
I'm not sure if I understand problem correctly, but seems for me it should work with nested collections. Please check test I added and provide more info if I got it wrong.

kobaeugenea replied:
Done

kobaeugenea commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/mapping/schema/ForeignKey.java:
```diff
@@ -0,0 +1,27 @@
+record ForeignKey(String name, String tableName, String columnName, String referencedTableName,
+									String referencedColumnName) {
+	@Override
+	public boolean equals(Object o) {
```
True, we don't need it. Deleted.

kobaeugenea commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/mapping/schema/SchemaDiff.java:
```diff
@@ -91,43 +92,40 @@ private static List<TableDiff> diffTable(Tables mappedEntities, Map<String, Tabl
-
+			tableDiff.columnsToDrop().addAll(findDiffs(mappedColumns, existingColumns, nameComparator));
+			// Identify added columns and add columns in order. This order can interleave with existing columns.
+			List<Column> addedColumns = new ArrayList<>(findDiffs(existingColumns, mappedColumns, nameComparator));
```
Make it to be `Collection`.

kobaeugenea commented:
Hey! I did some polishing and added test for nested collections. Please give me more information if I got your comment wrong.

schauder commented:
I'm sorry, I wasn't very precise when I said nested collections.

What I really should have said is nested lists or maps where the entities don't have ids.

I modified the domain model of your test to clarify what I mean: https://github.com/spring-projects/spring-data-relational/tree/pr/schema-generation-fks

dadmin-admin commented:
> I'm sorry, I wasn't very precise when I said nested collections.
> 
> What I really should have said is nested lists or maps where the entities don't have ids.
> 
> I modified the domain model of your test to clarify what I mean: https://github.com/spring-projects/spring-data-relational/tree/pr/schema-generation-fks

@schauder Oh, sorry don't quite understnad that case. I thought that every entity (class marked by Table annotation) must have field with Id annotation. If not I can't see where foreign key of OtherTable should reference in Tables. I can suggest that it can reference to composite key using all fields of Tables (then it makes sense to use collection of column names). But in that case why do you talk only about Map and List but not about Set?
Diffrence I know about Map/List and Set is MappedCollection for Map/List must have 'keyColumn' but it is for ordering as I understand and don't do anything with primary keys/foreign keys.

Sorry if my questions are stupid, just something doesn't match for me =)

kobaeugenea commented:
> I'm sorry, I wasn't very precise when I said nested collections.
> 
> What I really should have said is nested lists or maps where the entities don't have ids.
> 
> I modified the domain model of your test to clarify what I mean: https://github.com/spring-projects/spring-data-relational/tree/pr/schema-generation-fks

@schauder Oh, sorry don't quite understnad that case. I thought that every entity (class marked by Table annotation) must have field with Id annotation. If not I can't see where foreign key of OtherTable should reference in Tables. I can suggest that it can reference to composite key using all fields of Tables (then it makes sense to use collection of column names). But in that case why do you talk only about Map and List but not about Set?
Diffrence I know about Map/List and Set is MappedCollection for Map/List must have 'keyColumn' but it is for ordering as I understand and don't do anything with primary keys/foreign keys.

Sorry if my questions are stupid, just something doesn't match for me =)

schauder commented:
Not at all stupid questions. Inner entities of 1:1 relations or in lists or maps don't need ids. I actually recommend against ids.
The id of the aggregate root + the keycolumns for the lists/maps form a primary key. 

You can see examples of this in 
https://github.com/spring-projects/spring-data-relational/blob/7f649c4b66ac266ea4564f9215da67ae9f582336/spring-data-jdbc/src/test/java/org/springframework/data/jdbc/core/AbstractJdbcAggregateTemplateIntegrationTests.java#L879
And the matching schema in https://github.com/spring-projects/spring-data-relational/blob/7f649c4b66ac266ea4564f9215da67ae9f582336/spring-data-jdbc/src/test/resources/org.springframework.data.jdbc.core/JdbcAggregateTemplateIntegrationTests-hsql.sql#L149

kobaeugenea commented:
@schauder Oh, okay, now I see. Please check my last commit, I fixed it.

Just one questions to ensure. Is my guess right, that if parent entity already have composite primary key than "idColumn" of MappedCollection doesn't work and child entity reuses names from parent without changing (only adding keyColumn to primary key)?

Please look at top level entity (ListOfMapOfNoIdTables) in tests. I used "idColumn" there. Can you please check that primary key column names are correct in MapOfNoIdTables and NoIdTable based on that annotation.

kobaeugenea commented:
@schauder Hey! In last commit I added processing one-to-one relations, also simplify code a little, fix bugs and added few tricky test cases.
Please take a look on
`createForeignKeyForOneToOneWithMultipleChildren`
`createForeignKeyForCircularWithId`
`createForeignKeyForCircularNoId`
test cases. Escpecially I'm confused about this kind of entities
```
       @org.springframework.data.relational.core.mapping.Table
	static class OneToOneLevel2 {
		NoIdTable table1;
		@Column("additional_one_to_one_level2")
		NoIdTable table2;
	}
```
What is primary key in NoIdTable? one_to_one_level1 or additional_one_to_one_level2 or both or none? Because primary key is not nullable but what if one of the referencing fields will be null? Now I just pick first one available and make it primary key. Maybe we don't need to create primary key at all in such cases?

Here is even more interesting example:
```
       @org.springframework.data.relational.core.mapping.Table
	static class ParentOfCircularNoId {
		@Id int id;
		CircularNoId CircularNoId;

	}

	@org.springframework.data.relational.core.mapping.Table
	static class CircularNoId {
		CircularNoId CircularNoId;
	}
```
What schema should be generated here?

schauder commented:
If there are multiple child references of the same type and that type doesn't have an id, we strictly speaking shouldn't create PK. But I would be willing to fix this with a later change.

schauder commented:
Circular references aren't allowed and should fail, preferable with an expressive exception.

schauder commented:
I merged this after removing the tests for circular references, since these are not supported anyway.

Thanks.
